### ISSUE TITLE
`CCL_ZE_IPC_EXCHANGE` changed to sockets, crash with simple example

### ISSUE THREAD

**Issue by BenBrock** [intel/torch-ccl#71]
I'm trying to get a simple example working with PyTorch/CCL and keep running into crashes.

I'm running on a system with 4 Intel GPU Max 1550 GPUs (so 8 PVC tiles).  I installed IPEX/CCL with Conda via the [instructions detailed here](https://www.intel.com/content/www/us/en/developer/tools/oneapi/ai-tools-selector.html).  "AI Tools, Customize, Conda, Python 3.10, Radial button for Intel Extensions for PyTorch (GPU) selected."

I want to test out a simple program using P2P and some collectives with PyTorch dist using the CCL backend:

```python
#!/usr/bin/env python

import os
import torch
import torch.distributed as dist
import intel_extension_for_pytorch as ipex
import oneccl_bindings_for_pytorch as torch_ccl

def get_device():
    return 'xpu:%s' % (dist.get_rank() % torch.xpu.device_count(),)

os.environ["MASTER_ADDR"] = "127.0.0.1"
os.environ["MASTER_PORT"] = "29500"
os.environ["RANK"] = os.environ.get("PMI_RANK")
os.environ["WORLD_SIZE"] = os.environ.get("PMI_SIZE")
dist.init_process_group(backend="ccl", init_method="env://")

n = 4*1024*1024 // 4
nbytes = n * 4
gibytes = nbytes / 1024 / 1024 / 1024
tensor = torch.zeros(n, dtype=torch.float32, device=get_device())

dist.all_reduce(tensor)
```

When I launch this with MPI, I get crashes:

```bash
(ipex) bbrock@sdp562928:~/src/ai/pytorch-doodles/ccl-simple$ !mpiru
mpirun -n 8 ./simplest.py
2024:09:17-13:29:22:(146553) |CCL_WARN| value of CCL_ZE_IPC_EXCHANGE changed to be sockets (default:drmfd)
2024:09:17-13:29:24:(146553) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (0)
2024:09:17-13:29:24:(146553) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (0)
2024:09:17-13:29:24:(146554) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (1)
2024:09:17-13:29:24:(146554) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (1)
2024:09:17-13:29:24:(146555) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (2)
2024:09:17-13:29:24:(146555) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (2)
2024:09:17-13:29:24:(146556) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (3)
2024:09:17-13:29:24:(146556) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (3)
2024:09:17-13:29:24:(146557) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (4)
2024:09:17-13:29:24:(146557) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (4)
2024:09:17-13:29:24:(146558) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (5)
2024:09:17-13:29:24:(146558) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (5)
2024:09:17-13:29:24:(146559) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (6)
2024:09:17-13:29:24:(146559) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (6)
2024:09:17-13:29:24:(146560) |CCL_WARN| comm_dev_uuids is not sub-vector of node_dev_uuids, comm_dev_uuids size 8, node_dev_uuids size 1, this may happen due to narrow device affinity mask (7)
2024:09:17-13:29:24:(146560) |CCL_WARN| number of result device uuids does not match number of ranks per host, result size 1, host_rank_info_vec size 8, this may happen due to narrow device affinity mask (7)

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   RANK 4 PID 146557 RUNNING AT sdp562928
=   KILLED BY SIGNAL: 9 (Killed)
===================================================================================
```

The affinity mask warnings I think can be safely ignored---I am using MPI, and the default behavior has each MPI process observing only one PVC tile.

The `CCL_ZE_IPC_EXCHANGE` error, however, looks like CCL is not properly using Xe Link, and that might be causing the crashes?  Are there environment variables I need to be using to tell CCL to use Xe Link?  My PVCs are all fully connected with Xe Link.

**Comment by BenBrock** [comment]
This particular issue seems related to the default behavior of Intel MPI, which splits devices among processes.  Here, each process sees only a single device.  It seems like CCL is expecting to see _all_ devices within the node.  To make all the devices visible to every process, you can set `ZE_AFFINITY_MASK=0,1,2,3,4,5,6,7` (or the total number of devices available on your particular system) as [detailed here](https://spec.oneapi.io/level-zero/latest/core/PROG.html#environment-variables).

I also had to unset `CCL_ZE_IPC_EXCHANGE`, which was set by default to `sockets` with the default Conda install.

I also got warnings that PCIe was being used, and it seems setting `CCL_TOPO_FABRIC_VERTEX_CONNECTION_CHECK=0` will force CCL to assume that Xe Link is available (which is is on my system).

I am still encountering a couple of other issues, but I will open a new issue for that.

**Comment by BenBrock** [comment]
I've opened a new issue here (#73) for intermittent seg faults that I'm still observing with this configuration.
